### PR TITLE
Read env variables from KfDef Parameters

### LINKED ISSUE #179: Allow opendatahub-operator to read parameters from ENV variables
Extend `kfdef.spec.applications.kustomizeConfig.parameters.value` field to read values from environment variables.
For example:

```
  - kustomizeConfig:
      parameters:
        - name: s3_endpoint_url
          value: ${S3_ENDPOINT_URL}
```
 
Support `$VAR` and `${VAR}` syntax.

### PR DESCRIPTION
**Pull request by VaishnaviHire**
Fixes #179 

## How Has This Been Tested?
1. Install Open Data Hub operator
2. Update CSV with following two changes:
     - spec.image : quay.io/vhire/opendatahub-operator:env 
     - Add following env variable under spec.env
       ```
       name: MINIMAL_IMAGE
       value: quay.io/thoth-station/s2i-minimal-py38-notebook:v0.3.0
      ``` 
3. Deploy following KfDef
```

apiVersion: kfdef.apps.kubeflow.org/v1
kind: KfDef
metadata:
  name: opendatahub
  namespace: opendatahub
spec:
  applications:
  - kustomizeConfig:
      repoRef:
        name: manifests
        path: odh-common
    name: odh-common
  - kustomizeConfig:
      overlays:
      - additional
      parameters:
       - name: minimal-image
         value: $MINIMAL_IMAGE
      repoRef:
        name: manifests
        path: jupyterhub/notebook-images
    name: notebook-images
  repos:
  - name: manifests
    uri: https://github.com/VaishnaviHire/odh-manifests/tarball/test_image_vars
```
 

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Review by samuelvl — CHANGES_REQUESTED** [review]
It works fine if I define the parameters as following:

```yaml
parameters:
  - name: minimal-image
    value: $MINIMAL_IMAGE
```

Or

```yaml
parameters:
  - name: minimal-image
    value: quay.io/modh/odh-minimal-notebook-container:v1.18.0-2
```

However when using ${VAR} it doesn't work:

```yaml
parameters:
  - name: minimal-image
    value: ${MINIMAL_IMAGE}
```

I see the following error in the operator logs:

```shell
time="2022-10-19T11:28:35Z" level=warning msg="Encountered error applying application notebook-images:  (kubeflow.error): Code 500 with message: Apply.Run : ImageStream.image.openshift.io \"s2i-minimal-notebook\" is invalid: [[]: Internal error: imagestreams \"s2i-minimal-notebook\" is invalid: spec.tags[v0.3.0-py38].from.name: Invalid value: \"\": must be of the form <tag>, <repo>:<tag>, <id>, or <repo>@<id>, spec.tags[v0.3.0-py38].from.name: Required value]"
time="2022-10-19T11:28:35Z" level=warning msg="Will retry in 4 seconds."
``` 

I'm missing some unit tests to cover the `getParameterValue` function.

**Comment by Jooho** [comment]
@VaishnaviHire  keep in mind when the value is specified like the following:
~~~
parameters:
  - name: minimal-image
    value: ${MINIMAL_IMAGE}
~~~
In the pod, it will be changed like the following
~~~
parameters:
  - name: minimal-image
    value: '${MINIMAL_IMAGE}'
~~~
I am not sure your code will be impacted by this or not but I just want you to head up this kind openshift manifest behavior.

**Comment by VaishnaviHire** [comment]
> @VaishnaviHire keep in mind when the value is specified like the following:
> 
> ```
> parameters:
>   - name: minimal-image
>     value: ${MINIMAL_IMAGE}
> ```
> 
> In the kfdef object, it will be changed like the following
> 
> ```
> parameters:
>   - name: minimal-image
>     value: '${MINIMAL_IMAGE}'
> ```
> 
> I am not sure your code will be impacted by this or not but I just want you to head up this kind openshift manifest behavior.

Thanks ! This should not impact the code.

**Comment by VaishnaviHire** [comment]
@samuelvl Addressed the comments and updated the operator image

To run unit tests, use the following command
```
go test -run  TestGetParameterValue ./ -v
```

File: pkg/kfapp/kustomize/kustomize.go [hunk 1]
```diff
@@ -1462,3 +1463,25 @@ func GenerateYamlWithOperatorAnnotation(resMap resmap.ResMap, instance *unstruct
...
+	var re = regexp.MustCompile(`(?m)^\$({(.*)}|\w)`)
+	envVariable := ""
+	if re.Match([]byte(value)){
+		if strings.HasPrefix(value, "${"){
+			envVariable = value[2:len(value)-1]
+		}else if  strings.HasPrefix(value, "$") {
+			envVariable = value[1:]
+		}else {
+			return value
+		}
```
Comment (samuelvl) [inline pkg/kfapp/kustomize/kustomize.go:1479]: It works great now! Only one question, is this `return value` really needed?

```suggestion
		}
``` 

Running the coverage of this function I found that if I remove it we can achieve 100%

```shell
# With return value
$ go test -run TestGetParameterValue -v -covermode=count -coverprofile coverage ./pkg/kfapp/kustomize/...
$ go tool cover -func=coverage
github.com/kubeflow/kfctl/v3/pkg/kfapp/kustomize/kustomize.go:1467:	getParameterValue			90.9%

# After applying the suggestion
$ go test -run TestGetParameterValue -v -covermode=count -coverprofile coverage ./pkg/kfapp/kustomize/...
$ go tool cover -func=coverage
github.com/kubeflow/kfctl/v3/pkg/kfapp/kustomize/kustomize.go:1467:	getParameterValue			100.0%
```

**Review by samuelvl — APPROVED** [review]
/lgtm

**Comment by samuelvl** [comment]
@LaVLaS Could you approve this PR? :)

**Comment by VaishnaviHire** [comment]
@LaVLaS Did you get a chance to review this ? I think this can be merged for 1.4.1

**Pull request merged**